aemcmc/gibbs.py

```diff
@@ -13,52 +18,143 @@
+
+halfcauchy_1_lv, halfcauchy_2_lv = var(), var()
+zero_lv = var()
+horseshoe_pattern = etuple(
+    etuplize(at.random.normal),
+    var(),
+    var(),
+    var(),
+    zero_lv,
+    etuple(etuplize(at.mul), halfcauchy_1_lv, halfcauchy_2_lv),
+)
+
+
+def horseshoe_model(srng: TensorVariable) -> TensorVariable:
+    """Horseshoe shrinkage prior [1]_.
+
+    References
+    ----------
+    .. [1]: Piironen, J., & Vehtari, A. (2017). Sparsity information
+            and regularization in the horseshoe and other shrinkage priors.
+            Electronic Journal of Statistics, 11(2), 5018-5051.
+
+    """
+    size = at.scalar("size", dtype="int32")
+    tau_rv = srng.halfcauchy(0, 1, size=1)
+    lmbda_rv = srng.halfcauchy(0, 1, size=size)
+    beta_rv = srng.normal(0, tau_rv * lmbda_rv, size=size)
+    return beta_rv
+
+
+def horseshoe_match(graph: TensorVariable) -> Tuple[TensorVariable, TensorVariable]:
+    graph_opt = optimize_graph(graph)
+    graph_et = etuplize(graph_opt)
+
+    s = unify(graph_et, horseshoe_pattern)
+    if s is False:
+        raise ValueError("Not a horseshoe prior.")
+
+    halfcauchy_1 = eval_if_etuple(s[halfcauchy_1_lv])
+    if halfcauchy_1.owner is None or not isinstance(
+        halfcauchy_1.owner.op, type(at.random.halfcauchy)
+    ):
+        raise ValueError(
+            "Not a horseshoe prior. One of the shrinkage parameters "
+            + "in your model is not half-Cauchy distributed."
+        )
+
+    halfcauchy_2 = eval_if_etuple(s[halfcauchy_2_lv])
+
+    if halfcauchy_2.owner is None or not isinstance(
+        halfcauchy_2.owner.op, type(at.random.halfcauchy)
+    ):
+        raise ValueError(
+            "Not a horseshoe prior. One of the shrinkage parameters "
+            + "in your model is not half-Cauchy distributed."
+        )
+
+    if halfcauchy_1.type.shape == (1,):
+        lmbda_rv = halfcauchy_2
+        tau_rv = halfcauchy_1
+    elif halfcauchy_2.type.shape == (1,):
+        lmbda_rv = halfcauchy_1
+        tau_rv = halfcauchy_2
+    else:
+        raise ValueError(
+            "Not a horseshoe prior. The global shrinkage parameter "
+            + "in your model must be one-dimensional."
+        )
+
+    return (lmbda_rv, tau_rv)
+
+
 def horseshoe_step(
     srng: RandomStream,
     beta: TensorVariable,
-    sigma2: TensorVariable,
-    lambda2_inv: TensorVariable,
-    tau2_inv: TensorVariable,
+    sigma: TensorVariable,
+    lmbda_inv: TensorVariable,
+    tau_inv: TensorVariable,
 ) -> Tuple[TensorVariable, TensorVariable]:
-    r"""Gibbs kernel to sample from the posterior distribution of a horsehoe prior.
+    r"""Gibbs kernel to sample from the posterior distribution of a horseshoe prior.
 
     This kernel generates samples from the posterior distribution of the local
-    and global shrinkage parameters of a horsehoe prior, respectively :math:`\lambda`
+    and global shrinkage parameters of a horseshoe prior, respectively :math:`\lmbda`
```

Comment:
```suggestion
    and global shrinkage parameters of a horseshoe prior, respectively :math:`\lambda`
```
In this case, will the change to `\lmbda` even render in LaTeX?  I don't recall an alias like that.

Reply:
Good catch